maven-caching/src/main/java/org/apache/maven/caching/checksum/MavenProjectInput.java

```diff
@@ -743,75 +695,49 @@ private boolean isFilteredOutSubpath( Path path )
             }
 
             // saved to index by the end of dependency build
-            final boolean currentlyBuilding = isBuilding( dependency );
-            final boolean partOfMultiModule = strategy.isPartOfMultiModule( dependency );
-            if ( !currentlyBuilding && !partOfMultiModule && !isSnapshot( dependency.getVersion() ) )
+            MavenProject dependencyProject = multiModuleSupport.tryToResolveProject(
+                            dependency.getGroupId(),
+                            dependency.getArtifactId(),
+                            dependency.getVersion() )
+                    .orElse( null );
+            boolean isSnapshot = isSnapshot( dependency.getVersion() );
+            if ( dependencyProject == null && !isSnapshot )
             {
                 // external immutable dependency, should skip
                 continue;
             }
-
-            final Artifact dependencyArtifact = repoSystem.createDependencyArtifact( dependency );
-            final String artifactKey = KeyUtils.getArtifactKey( dependencyArtifact );
-            DigestItem resolved = null;
-            if ( currentlyBuilding )
+            String projectHash;
+            if ( dependencyProject != null ) //part of multi module
             {
-                resolved = projectArtifactsByKey.get( artifactKey );
-                if ( resolved == null )
-                {
-                    throw new DependencyNotResolvedException( "Expected dependency not resolved: " + dependency );
-                }
+                projectHash = projectInputCalculator.calculateInput( dependencyProject ).getChecksum();
```

Comment:
result is not cached, potentially could be recalculated many times for the same dependency in different modules

Reply:
> 

the result is cached take a look at https://github.com/apache/maven/pull/623/files#diff-e4da53c029acd1f06075465c790bd5c3f8330048ee490be33cac19d49a5b9130R53